### PR TITLE
Disable Default value assignment for columns with Dynamic Defaults

### LINKED ISSUE #956: Default ID column's dynamic default can be altered
## Description

As noted in [this comment](https://github.com/centerofci/mathesar/pull/941#pullrequestreview-844879976) by @kgodey on #941 , it's currently possible to alter default columns via the API through `PATCH` requests.

## Expected behavior

It should be impossible to change the default generator of the default `id` column via the API.  In particular, if the user alters the default value, they'll have trouble since the column is required to remain unique, and they probably won't want to pick an unused integer for every row.

## To Reproduce

Give a `PATCH` request that to the appropriate columns endpoint that alters the default value of the `id` column.  The form of this request is currently in flux, so we will leave specifying it for a bit later.

## Additional context

Should it be possible to alter other attributes of the default `id` column, e.g., the name?  @kgodey @ghislaineguerin I'd appreciate a quick weigh-in on that.

### PR DESCRIPTION
**Pull request by Aritra8438**
This PR is related to #956 

Currently, Dynamic Defaults of columns can be altered and static defaults can be assigned to the same. 

## Technical details

This PR modifies 
- validation of the `PATCH` request such that, dynamic defaults of a column can't be altered.
- Add test for columns with dynamic default

**Screencasts**

https://user-images.githubusercontent.com/64671908/229990085-d475f8fa-ad72-4b71-ad0e-fc4b87d82852.mp4

## Checklist
 
- [x] My pull request has a descriptive title (not a vague title like `Update
  index.md`).
- [x] My pull request targets the `develop` branch of the repository
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [x] I added tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no
  visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

**Review by silentninja — CHANGES_REQUESTED** [review]
Thanks for the PR! @Aritra8438. The PR addresses the problem correctly but there are some idiomatic changes that need to be fixed before the PR can be merged.

Also, I am not entirely sure if the PR addresses #956 as the description of #956 is quite confusing and I can see that you have made [comments](https://github.com/centerofci/mathesar/issues/956#issuecomment-1495352873) with regards to that . Until there is an answer to your comment, please change the description of the PR to say that it is `Related to #956` and not `Fixes #956`

File: mathesar/utils/columns.py
Comment (Aritra8438) [reply mathesar/utils/columns.py:5]: Sure, @silentninja, I will fix that.

File: mathesar/api/db/viewsets/columns.py
Comment (Aritra8438) [reply mathesar/api/db/viewsets/columns.py:125]: > so the `PATCH` request should throw a validation error if the `default` field is present.

As **Kriti mentioned in this [comment](https://github.com/centerofci/mathesar/issues/956#issuecomment-1505269803)**,

- **If a column has dynamic default**, the patch request should not be able to modify the default of that column.

**Primary key column has dynamic default**. As you have mentioned, the fix will be **not to allow default field in the patch request if a column has dynamic default**. I will update the PR accordingly.

**Comment by Aritra8438** [comment]
Hi, @silentninja, I have done the necessary changes and the exception is raised successfully.
But, **if only patch request has a default value, the validation check is throwing another error:**

```bash
mathesar_service_dev   | Internal Server Error: /api/db/v0/tables/499/columns/911/
mathesar_service_dev   | Traceback (most recent call last):
mathesar_service_dev   |   File "/usr/local/lib/python3.9/site-packages/django/core/handlers/exception.py", line 47, in inner
mathesar_service_dev   |     response = get_response(request)
mathesar_service_dev   |   File "/usr/local/lib/python3.9/site-packages/django/core/handlers/base.py", line 204, in _get_response
mathesar_service_dev   |     response = response.render()
mathesar_service_dev   |   File "/usr/local/lib/python3.9/site-packages/django/template/response.py", line 105, in render
mathesar_service_dev   |     self.content = self.rendered_content
mathesar_service_dev   |   File "/usr/local/lib/python3.9/site-packages/rest_framework/response.py", line 70, in rendered_content
mathesar_service_dev   |     ret = renderer.render(self.data, accepted_media_type, context)
mathesar_service_dev   |   File "/usr/local/lib/python3.9/site-packages/rest_framework/renderers.py", line 724, in render
mathesar_service_dev   |     context = self.get_context(data, accepted_media_type, renderer_context)
mathesar_service_dev   |   File "/usr/local/lib/python3.9/site-packages/rest_framework/renderers.py", line 702, in get_context
mathesar_service_dev   |     'filter_form': self.get_filter_form(data, view, request),
mathesar_service_dev   |   File "/usr/local/lib/python3.9/site-packages/rest_framework/renderers.py", line 634, in get_filter_form
mathesar_service_dev   |     html = backend().to_html(request, queryset, view)
mathesar_service_dev   |   File "/usr/local/lib/python3.9/site-packages/rest_framework/filters.py", line 304, in to_html
mathesar_service_dev   |     context = self.get_template_context(request, queryset, view)
mathesar_service_dev   |   File "/usr/local/lib/python3.9/site-packages/rest_framework/filters.py", line 296, in get_template_context
mathesar_service_dev   |     for key, label in self.get_valid_fields(queryset, view, context):
mathesar_service_dev   |   File "/usr/local/lib/python3.9/site-packages/rest_framework/filters.py", line 250, in get_valid_fields
mathesar_service_dev   |     return self.get_default_valid_fields(queryset, view, context)
mathesar_service_dev   |   File "/usr/local/lib/python3.9/site-packages/rest_framework/filters.py", line 230, in get_default_valid_fields
mathesar_service_dev   |     model_property_names = [
mathesar_service_dev   |   File "/usr/local/lib/python3.9/site-packages/rest_framework/filters.py", line 232, in <listcomp>
mathesar_service_dev   |     attr for attr in dir(model_class) if isinstance(getattr(model_class, attr), property) and attr != 'pk'
mathesar_service_dev   |   File "/code/mathesar/state/cached_property.py", line 56, in __get__
mathesar_service_dev   |     key = self._get_ip_key(instance=instance)
mathesar_service_dev   |   File "/code/mathesar/state/cached_property.py", line 82, in _get_ip_key
mathesar_service_dev   |     ip_key = self._get_random_ip_key(instance)
mathesar_service_dev   |   File "/code/mathesar/state/cached_property.py", line 111, in _get_random_ip_key
mathesar_service_dev   |     ip_key = self._get_ip_key_from_instance_cache(instance)
mathesar_service_dev   |   File "/code/mathesar/state/cached_property.py", line 124, in _get_ip_key_from_instance_cache
mathesar_service_dev   |     instance_cache = self._get_instance_cache(instance)
mathesar_service_dev   |   File "/code/mathesar/state/cached_property.py", line 138, in _get_instance_cache
mathesar_service_dev   |     return instance.__dict__
mathesar_service_dev   | AttributeError: 'NoneType' object has no attribute '__dict__'
```

The AttributeError refers to `_get_instance_cache` which is receiving a **NoneType** object **when the exception is raised**. 

Can you take a look?

**Comment by Aritra8438** [comment]
Hi, @silentninja, should I modify anything else?

**Comment by silentninja** [comment]
> Hi, @silentninja, should I modify anything else?

Sorry for holding this PR, I am a little busy with the v0.1.2 milestone issues. I will look into this PR tomorrow(Tuesday)

**Review by silentninja — CHANGES_REQUESTED** [review]
Thanks for addressing the requests. However, there are a few minor issues which I have outlined in specific comments. 
Also, your PR title does not match the description. Please make changes to the PR and its description if you are only planning to throw an error when setting static defaults for the primary key column

File: mathesar/tests/api/test_column_api.py
Comment (Aritra8438) [reply mathesar/tests/api/test_column_api.py:434]: **As this PR restricts default for primary_key column,** I'm not sure the only column tested here is primary_key column or not.
**If it is, then what will be the solution?** (As far as I remember, it was.)

**Comment by Aritra8438** [comment]
Hi, @silentninja, as mentioned in the issue #956, **this PR only restricts the default modification of a column when it already has a dynamic default.** 
As, primary_key columns are only such columns, **I am modifying the PR name based on the primary_key column.**

**Comment by silentninja** [comment]
> As, primary_key columns are only such columns, I am modifying the PR name based on the primary_key column.

We also reflect columns which were not created using Mathesar. Such columns can have dynamic defaults, so we need to account for such scenario and not restrict it to only primary key column

File: mathesar/tests/api/test_column_api.py
Comment (silentninja) [reply mathesar/tests/api/test_column_api.py:434]: The test case is not wrong by any means, it is just that it is testing too many things. It is easier to understand the test case if it asserts only one behaviour at a time. You should add a new test just for testing the primary key column 

- Keep the old test as it is and create a new test called `test_pk_column_update_delete_default` and test only the behaviour of the primary key column.

```diff
def test_column_update_delete_default(column_test_table, client):
    ...
    if is_primary_column(column_id=column.id, table=column_test_table):
          assert response.json()[0]["code"] == 4207
      else:
          assert response.json()["default"]["value"] == expt_default
-     else:
-        assert response.json()["default"]["value"] == expt_default
```

**Comment by Aritra8438** [comment]
> We also reflect columns which were not created using Mathesar. Such columns can have dynamic defaults, so we need to account for such scenario and not restrict it to only primary key column

Hi, @silentninja, I wasn't aware of this, thanks for letting me know. 

So, I modified the PR to **restrict the ```PATCH``` request for columns with Dynamic default** as mentioned by Brent here. 

https://github.com/centerofci/mathesar/issues/956#issue-1095040819 - ```It should be impossible to change the default generator of a column via the API, until we have functionality that would enable the user to recover from doing so in error.```

**Comment by silentninja** [comment]
@Aritra8438 Sorry for taking too long to review the PR. I will review it today.

**Comment by silentninja** [comment]
@Aritra8438 I made a few changes before merging the PR

- The `DynamicDefaultModificationError` has been moved into the `mathesar` module because the exception is specific to Mathesar App(Mathesar cannot handle creating dynamic defaults **now**) and is not related to the Database.
- The superfluous serializer wrapper(`instance_serializer`) when validating the column data is removed as the`column_default_dict` field can be accessed directly.

**Comment by Aritra8438** [comment]
Thanks, @silentninja.

As per my understanding,

`DynamicDefaultModificationError` is not related to the database (as it is possible to do so) and we might want to add support for modifying the dynamic default later. Until then, it should be classified as Mathesar api exception.

**Review by silentninja — APPROVED** [review]

**Pull request merged**
